### PR TITLE
refactor(sqlengine): add extensible filter set rewriting

### PR DESCRIPTION
**Pull request by egolearner**
## Summary

- Centralize recursive filter normalization and introduce registries for group and set rewrite rules.
- Fix the incorrect rewrite of OR-connected not-equal predicates into NOT IN.
- Merge same-field equality/IN predicates under OR and inequality/NOT IN predicates under AND.
- Deduplicate exact literals in merged results using string_view keys while leaving standalone IN/NOT IN lists unchanged.
- Keep positive intersections under AND and semantic value canonicalization outside the initial rule set.
- Add regression tests for rewrite boundaries, invalid literals, duplicate handling, predicate ordering, and unsatisfiable-filter propagation.

## Testing

- ninja -C build.release unittest.zvec_sqlengine
- clang-format -n --Werror on changed C++ files
- git diff --check

### CODE REVIEW

File: src/db/sqlengine/analyzer/simple_rewriter.cc
Comment (zhourrr) [inline src/db/sqlengine/analyzer/simple_rewriter.cc:1]: The forward-scan path still incorrectly includes NULL rows in NOT IN results because of Arrow’s default SetLookupOptions::MATCH. The new != AND != → NOT IN rewrite also exposes this issue. The fix is to use SetLookupOptions::INCONCLUSIVE, with regression tests covering nullable fields with and without indexes.

Comment (egolearner) [reply src/db/sqlengine/analyzer/simple_rewriter.cc:1]: Fixed in 8dde0049. The forward-scan IN/NOT IN expression now uses SetLookupOptions::INCONCLUSIVE, so NULL inputs stay unknown instead of being selected by NOT IN. I added regression tests for nullable fields with and without inverted indexes, covering direct NOT IN and the != AND != rewrite. Those tests also exposed a separate issue: a merged predicate promoted to the filter root retained the old logic node as its parent, allowing the analyzer to drop an indexed filter. The rewritten root now clears that stale parent pointer. The full unittest.zvec_sqlengine target passes.

File: src/db/sqlengine/analyzer/simple_rewriter.cc [hunk 1]
```diff
@@ -13,194 +13,507 @@
...
+      return {std::move(node), TruthValue::ALWAYS_FALSE, true};
+    }
+    if (left.truth == TruthValue::ALWAYS_FALSE) {
+      if (right.truth != TruthValue::DYNAMIC) {
+        return {std::move(right.root), right.truth, true};
+      }
+      return reduce_group_if_needed(std::move(right.root), schema, reduce_group,
+                                    true);
+    }
+    if (right.truth == TruthValue::ALWAYS_FALSE) {
```
Comment (zhourrr) [inline src/db/sqlengine/analyzer/simple_rewriter.cc:468]: 当前重写先于字段和类型校验，可能把非法条件提前删除。例如 `category_set CONTAIN_ALL () OR missing_field CONTAIN_ANY ()`，其中 `missing_field` 不存在，PR 前会报错，当前却会将右侧判为恒假并删除，最终接受查询。仅检查空 CONTAIN 自身的字段还不够：在 `age=1 OR (category_set CONTAIN_ANY () AND missing_field=2)` 中，非法条件也会随着整个恒假的 AND 子树被删除。建议在重写前，对原始条件树执行一次不修改 AST 的字段、操作符和类型校验，可以给现有 SearchCondCheckWalker 增加只校验模式，禁止节点标记及数值转换结果的回写，重写后再执行现有正常分析流程。4096 个条件的数量限制仍保留在重写后，并补充嵌套条件中非法字段和错误类型被删除的回归测试。

File: src/db/sqlengine/analyzer/simple_rewriter.cc [hunk 2]
```diff
@@ -13,194 +13,507 @@
...
   }
-  if (rewrited) {
-    simplify_tree(query_node, query_info);
-    std::string after_rewrite = query_info->search_cond()->text();
-    LOG_INFO("Rewrite filter. before[%s] after[%s]", before_rewrite.c_str(),
-             after_rewrite.c_str());
+  QueryNode::Ptr root = std::move(terms.front());
+  for (size_t i = 1; i < terms.size(); ++i) {
+    auto parent = std::make_shared<QueryNode>(logic);
+    parent->set_left(std::move(root));
```
Comment (zhourrr) [inline src/db/sqlengine/analyzer/simple_rewriter.cc]: 当前 PR 的 `build_logic_tree()` 会把合并后的 OR 组重建成左深链，抵消 Parser 对长 OR 表达式已有的平衡处理。这个问题主要影响合并后仍保留大量条件的表达式，例如许多 `(tenant_id = … AND role = …)` 通过 OR 连接，末尾再接 `owner_id = 10 OR owner_id = 20`。两版都会把末尾条件合成 `owner_id IN (10, 20)`，并完整保留前面的 `AND` 组合，但旧版只替换相关节点、保留其余平衡结构，当前 PR 却会重建整个 OR 组。实际对照中，4 个 OR 项时两版树深度相同；256 项时，旧版深度为 11，PR 为 257；1024 项时，旧版为 13，PR 为 1025。因此，两版完成了相同的合并，当前实现却显著增加了长表达式的递归深度，可能增加栈消耗和子树字符串构造成本。目前已确认结构退化，尚未复现崩溃或量化性能损失。建议重建长 OR 组时沿用 Parser 的平衡策略，保持条件顺序及 AND 子树边界，并补充“长 OR 组中只有少量条件发生合并”的树深度回归测试。

Comment (egolearner) [reply src/db/sqlengine/analyzer/simple_rewriter.cc]: 已在 be9d5479 修复。规则命中后不再把整个逻辑组重建为左深链，而是沿原树骨架收缩 inactive 叶子，并仅在单侧为空时提升另一侧，因此保留 Parser 已有的平衡 OR 结构、条件顺序及 AND 子树边界。新增 1024 项部分可折叠 OR 测试：其中仅末尾两个条件合并，修复前深度为 1022，修复后不超过 12。unittest.zvec_sqlengine 已通过。

File: src/db/sqlengine/analyzer/simple_rewriter.cc (hunk 1 above)
Comment (egolearner) [reply src/db/sqlengine/analyzer/simple_rewriter.cc:468]: 已在 be9d5479 修复。重写前新增 VALIDATE_ONLY 模式遍历原始条件树：仍执行字段、操作符、类型和数值可转换性校验，但不写入 or_ancestor、vector/forward/invert 标记，也不回写数值 buffer；4096 条件限制仍在重写后的正常分析阶段执行。新增嵌套恒假分支中的缺失字段和错误类型回归测试，修复前均可复现被错误接受，修复后会正确返回错误。unittest.zvec_sqlengine 已通过。

**Review by zhourrr — APPROVED** [review]
